src/fields/field_value_cache.cc

```diff
@@ -0,0 +1,128 @@
+ */
+
+const unsigned int ElementCacheMap::n_cached_elements = 20;
+
+const unsigned int ElementCacheMap::undef_elem_idx = std::numeric_limits<unsigned int>::max();
+
+
+ElementCacheMap::ElementCacheMap()
+: elm_idx_(ElementCacheMap::n_cached_elements, ElementCacheMap::undef_elem_idx),
+  begin_idx_(0), end_idx_(0), dim_(ElementCacheMap::undef_elem_idx) {}
+
+
+void ElementCacheMap::add(const DHCellAccessor &dh_cell) {
+    ASSERT_LT(added_elements_.size(), ElementCacheMap::n_cached_elements).error("ElementCacheMap overflowed. List of added elements is to long!\n");
+    unsigned int elm_idx = dh_cell.elm_idx();
+   	added_elements_.insert(elm_idx);
+}
+
+
+void ElementCacheMap::prepare_elements_to_update() {
+    // Compute number of element stored in data cache (stored in previous cache update).
+    unsigned int n_stored_element = 0;
+    for (auto elm_idx : added_elements_)
+        if (cache_idx_.find(elm_idx) != cache_idx_.end()) ++n_stored_element;
+
+    // Test if new elements can be add the end of cache
+    if (cache_idx_.size() + added_elements_.size() - n_stored_element <= ElementCacheMap::n_cached_elements) {
+    	// Erase elements from added_elements set that exist in cache
+        for (auto it = added_elements_.begin(); it != added_elements_.end();) {
+            if ( cache_idx_.find(*it) != cache_idx_.end() ) it = added_elements_.erase(it);
+            else it++;
+        }
+        begin_idx_ = cache_idx_.size();
+    } else {
+    	// Clear cache, there is not sufficient space for new elements
+        std::fill(elm_idx_.begin(), elm_idx_.end(), ElementCacheMap::undef_elem_idx);
+        cache_idx_.clear();
+        begin_idx_ = 0;
+    }
+    end_idx_ = begin_idx_ + added_elements_.size();
+
+    // Add new elements indices to cache_idx_ and elm_idx_
+    unsigned int cache_pos = begin_idx_;
+    for (auto el_idx : added_elements_) {
+    	cache_idx_[el_idx] = cache_pos;
+		elm_idx_[cache_pos] = el_idx;
+        cache_pos++;
+    }
+    std::cout << "ElementCacheMap::prepare_elements_to_update: " << begin_idx_ << ", " << end_idx_ << std::endl;
+    for (auto i : added_elements_) std::cout << " - " << i;
+    std::cout << std::endl;
+}
+
+
+void ElementCacheMap::clear_elements_to_update() {
+	std::cout << "ElementCacheMap::clear_elements_to_update" << std::endl;
+	added_elements_.clear();
+}
+
+
+DHCellAccessor & ElementCacheMap::operator() (DHCellAccessor &dh_cell) const {
+	unsigned int elm_idx = dh_cell.elm_idx();
+	std::unordered_map<unsigned int, unsigned int>::const_iterator it = cache_idx_.find(elm_idx);
+	if ( it != cache_idx_.end() ) dh_cell.set_element_cache_index( it->second );
+	else dh_cell.set_element_cache_index( ElementCacheMap::undef_elem_idx );
+    return dh_cell;
+}
+
+
+/******************************************************************************
+ * Explicit instantiation of templates
+ */
+
+template class FieldValueCache<FieldValue<0>::Enum >;    // NOT tested, necessary for linking!
```

Comment:
Use Armor::Mat as template parameter instead of FieldValue:
- Armor::Mat<double, 1,1>::ArmaType (double)
- Armor::Mat<double, 3,1>::ArmaType (arma::vec3)
- Armor::Mat<double, 3,3>::ArmaType (arma::mat33)
- Armor::Mat<uint, 1,1>::ArmaType (uint)